"ciphertext verification failed" with version 0.8.2

## Output of `restic version`

```
restic 0.8.2
compiled with go1.10 on linux/amd64
```

## How did you run restic exactly?

```
docker run -it restic/restic:0.8.2 check
```

## What backend/server/service did you use to store the repository?

minio:RELEASE.2017-11-22T19-55-46Z

I have just tried the latest one: `minio:RELEASE.2018-02-09T22-40-05Z`.

## Expected behavior

It should returns something like:

```
password is correct
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
  signal interrupt received, cleaning up
read all data
...
no errors were found
```

## Actual behavior

```
ciphertext verification failed
github.com/restic/restic/internal/crypto.init
        /tmp/restic-build-354782844/src/github.com/restic/restic/internal/crypto/crypto.go:30
github.com/restic/restic/internal/restic.init
        <autogenerated>:1
github.com/restic/restic/internal/archiver.init
        <autogenerated>:1
main.init
        <autogenerated>:1
runtime.main
        /usr/local/go/src/runtime/proc.go:186
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:2361
```

## Steps to reproduce the behavior

- install minio-server
- run `restic check` with version from 0.8.2

## Do you have any idea what may have caused this?

It does not happen with version prior to 0.8.1 that I'm running `check --read-data` with. Output will be posted when it's done.

c4e9d5d11e27aa8d2d6523a1f38a15e46fbdebfb cannot be built:

```
❯❯❯❯ go run build.go -o restic.c4e9d5d11e27aa8d2d6523a1f38a15e46fbdebfb
# github.com/restic/restic/internal/backend/b2
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/b2/b2.go:31:5: cannot use b2Backend literal (type *b2Backend) as type restic.Backend in assignment:
	*b2Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/b2/b2.go:77:2: cannot use be (type *b2Backend) as type restic.Backend in return argument:
	*b2Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/b2/b2.go:127:2: cannot use be (type *b2Backend) as type restic.Backend in return argument:
	*b2Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
# github.com/restic/restic/internal/backend/rest
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/rest/rest.go:24:5: cannot use restBackend literal (type *restBackend) as type restic.Backend in assignment:
	*restBackend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/rest/rest.go:94:2: cannot use be (type *restBackend) as type restic.Backend in return argument:
	*restBackend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
# github.com/restic/restic/internal/backend/sftp
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/sftp/sftp.go:35:5: cannot use SFTP literal (type *SFTP) as type restic.Backend in assignment:
	*SFTP does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
# github.com/restic/restic/internal/backend/swift
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/swift/swift.go:33:5: cannot use beSwift literal (type *beSwift) as type restic.Backend in assignment:
	*beSwift does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/swift/swift.go:94:2: cannot use be (type *beSwift) as type restic.Backend in return argument:
	*beSwift does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
# github.com/restic/restic/internal/backend/azure
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/azure/azure.go:31:5: cannot use Backend literal (type *Backend) as type restic.Backend in assignment:
	*Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/azure/azure.go:67:2: cannot use *Backend as type restic.Backend in return argument:
	*Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/azure/azure.go:88:2: cannot use be (type *Backend) as type restic.Backend in return argument:
	*Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
# github.com/restic/restic/internal/repository
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/repository/key.go:120:14: cannot range over s.Backend().List(ctx, restic.KeyFile) (type error)
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/repository/key.go:120:36: not enough arguments in call to s.Backend().List
	have (context.Context, string)
	want (context.Context, restic.FileType, func(restic.FileInfo) error)
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/repository/parallel.go:24:17: not enough arguments in call to repo.List
	have (context.Context, restic.FileType)
	want (context.Context, restic.FileType, func(restic.FileInfo) error)
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/repository/parallel.go:34:17: select case must be receive, send or assign recv
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/repository/parallel.go:34:20: invalid operation: <-ch (receive from non-chan type error)
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/repository/repository.go:544:16: cannot range over r.be.List(ctx, t) (type error)
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/repository/repository.go:544:31: not enough arguments in call to r.be.List
	have (context.Context, restic.FileType)
	want (context.Context, restic.FileType, func(restic.FileInfo) error)
# github.com/restic/restic/internal/backend/gs
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/gs/gs.go:42:5: cannot use Backend literal (type *Backend) as type restic.Backend in assignment:
	*Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/gs/gs.go:99:2: cannot use *Backend as type restic.Backend in return argument:
	*Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/gs/gs.go:157:2: cannot use be (type *Backend) as type restic.Backend in return argument:
	*Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
# github.com/restic/restic/internal/backend/s3
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/s3/s3.go:32:5: cannot use Backend literal (type *Backend) as type restic.Backend in assignment:
	*Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/s3/s3.go:96:2: cannot use *Backend as type restic.Backend in return argument:
	*Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-882780403/src/github.com/restic/restic/internal/backend/s3/s3.go:120:2: cannot use be (type *Backend) as type restic.Backend in return argument:
	*Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
build failed: exit status 2
exit status 1
```

Same as 52230b8f0738c11444d70232905d1c1d32f7cb5e:

```
❯❯❯❯ go run build.go -o restic.52230b8f0738c11444d70232905d1c1d32f7cb5e
# github.com/restic/restic/internal/backend/b2
/var/folders/kn/b0zy3rkn1dz_5qpdx60vddp40000gn/T/restic-build-440791939/src/github.com/restic/restic/internal/backend/b2/b2.go:31:5: cannot use b2Backend literal (type *b2Backend) as type restic.Backend in assignment:
	*b2Backend does not implement restic.Backend (wrong type for List method)
		have List(context.Context, restic.FileType) <-chan string
		want List(context.Context, restic.FileType, func(restic.FileInfo) error) error
```

--- a/internal/repository/key.go
+++ b/internal/repository/key.go
@@ -115,8 +115,11 @@ func OpenKey(ctx context.Context, s *Repository, name string, password string) (
 func SearchKey(ctx context.Context, s *Repository, password string, maxKeys int) (k *Key, err error) {
 	checked := 0
 
-	// try at most maxKeysForSearch keys in repo
-	err = s.Backend().List(ctx, restic.KeyFile, func(fi restic.FileInfo) error {
+	listCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// try at most maxKeys keys in repo
+	err = s.Backend().List(listCtx, restic.KeyFile, func(fi restic.FileInfo) error {
 		if maxKeys > 0 && checked > maxKeys {
 			return ErrMaxKeysReached
 		}
@@ -142,8 +145,13 @@ func SearchKey(ctx context.Context, s *Repository, password string, maxKeys int)
 
 		debug.Log("successfully opened key %v", fi.Name)
 		k = key
+		cancel()
 		return nil
 	})
+
+	if err == context.Canceled {
+		err = nil
+	}
 
 	if err != nil {
 		return nil, err

